docs/modules/SDDs/pages/0025-commodore-component-instantiation.adoc

```diff
@@ -0,0 +1,100 @@
+
+=== Goals
+
+* Define how we want to allow Commodore components to be instantiated multiple times per cluster.
+
+=== Non-Goals
+
+* Define a solution for managing multiple instances of services for the customer on a Syn-enabled cluster.
+
+== Design Proposal
+
+Commodore will provide a well-defined structure for instantiating components and providing per-instance parameters.
+Commodore will provide the instance's identifier ("name") to the component in a well-defined way.
+Component authors must explicitly declare that their component supports instantiation.
+Similar to Helm charts, the components themselves must make sure to not cause any naming collisions of objects
+This is required both for namespaced and non-namespaced resources.
+By default, Commodore will place component instances in a single namespace.
+
+Rather than requiring users to manually implement "common" parameters, Commodore may provide support for defining "common" parameters which are used for all instances unless overridden.
+
+The astute reader will notice that the proposed design combines features from both "pure" approaches listed in <<Alternatives>>.
+
+=== User Stories
+
+* As a user, I would like to configure multiple instances of `nfs-client-provisioner` which provision volumes on different NFS shares on my Project Syn-managed clusters.
+  Ideally, I would like to be able to provide shared configuration options, such as mount options only once.
+
+== Alternatives
+
+=== Multi-instance aware components
+
+One alternative approach for supporting instantiating components, is to simply leave implementation to the component authors.
+A pure form of this approach needs no changes in Commodore, but has some significant drawbacks from a usability perspective.
+For this approach, component authors need to implement multi-instance support in each component individually.
+
+This approach works well for infrastructure tooling for which a Kubernetes operator exists.
+If an operator exists, the component can install the operator and associated CRDs, and can render instances as custom resources for the operator to provision.
+
+The main drawback of the pure approach is that it may not be feasible to implement without some changes in Commodore for components which use Helm charts as their base, unless the Helm chart exposes a similar structure.
+The reason for this is that Kapitan's Helm templating is expressed in the reclass inventory, and reclass isn't flexible enough to instantiate multiple copies of a part of the hierarchy based on another key in the inventory.
+
+While it's ultimately left to the component authors, we assume that many components will expose the multi-instance nature of the component in `parameters.<component-name>`.
+Taking `nfs-client-provisioner` as our working example, the parameters structure might look roughly as reproduced below.
```

Comment:
Another example:

```
applications:
  - nfs-client-provisioner
  - nfs-client-provisioner as nfs-server1
  - nfs-client-provisioner as nfs-server2

parameters:
  nfs_client_provisioner:
    namespace: syn-nfs-provisioner
    host: nfs.example.com
    path: /nfs
  nfs_server1:
    host: nfs1.example.com
  nfs_server2:
    host: nfs2.example.com
```

1. Using the `as` keyword in an application (component include) string instantiates the component with the defined alias.
2. Commodore will merge the dict `$alias` with `nfs_client_provisioner` to provide/override defaults.
3. Inside the component a helper function wraps `kap.inventory().parameters.$alias` to provide the correct values.

If no alias is provided, the component is imported with it's default name.

Reply:
>     3\. Inside the component a helper function wraps `kap.inventory().parameters.$alias` to provide the correct values.

This is not even necessary, as we can simply merge the alias config values into `nfs_client_provisioner` in the target for the aliased variant:

```
# inventory/targets/nfs-server1.yml
classes:
- [ ... defaults ... ]
- global.commodore
- components.nfs-client-provisioner

parameters:
  kapitan:
    vars:
      target: nfs-server1.yml
  nfs_client_provisioner: ${nfs_server1}

Reply:
Also note, that this suggestion does not work for the no Commodore support approach, and the original example is not what will be implemented. I'll update the "Design Proposal" section, with this example approach and my variant of it.